yarn-project/simulator/src/avm/avm_tree.ts

```diff
@@ -0,0 +1,808 @@
+    return result;
+  }
+
+  /**
+   * This hashes the preimage to a field element
+   */
+  hashPreimage<T extends TreeLeafPreimage>(preimage: T): Fr {
+    const input = preimage.toHashInputs().map(x => Fr.fromBuffer(x));
+    return poseidon2Hash(input);
+  }
+
+  /**
+   * This builds on the getLowInfo function to get the sibling path and also checks if this is an update or append operation
+   * @param treeId - The tree we are looking up
+   * @param value - The value for which we are look up the low leaf for, either a slot or a nullifier
+   * @param T - The type of the preimage (PublicData or Nullifier)
+   * @returns The low witness which contains the low sibling path, the low leaf preimage and the index of the low leaf in the indexed tree
+   */
+  private async _getIndexedLowWitness<ID extends IndexedTreeId, T extends IndexedTreeLeafPreimage>(
+    treeId: ID,
+    value: Fr,
+    T: { fromBuffer: (buffer: Buffer) => T },
+  ): Promise<LowWitness<T>> {
+    // Get low public data witness
+    const { preimage: lowWitness, index } = await this.getLowInfo(treeId, value, T);
+    // We check if the value is already in the tree
+    const isUpdate: boolean = lowWitness.getNextKey() === value.toBigInt();
+    if (isUpdate) {
+      const lowSiblingPath =
+        this.treeMap.get(treeId)!.getSiblingPath(index) ?? (await this.treeDb.getSiblingPath(treeId, index)).toFields();
+      const result = {
+        update: true,
+        lowSiblingPath: lowSiblingPath,
+        preimage: lowWitness,
+        index,
+      };
+      return result;
+    }
+    // This is an insertion of a new entry
+    const lowSiblingPath =
+      this.treeMap.get(treeId)!.getSiblingPath(index) ?? (await this.treeDb.getSiblingPath(treeId, index)).toFields();
+    const oldLowWitness = T.fromBuffer(lowWitness.toBuffer());
+
+    // Build Result struct
+    const result = {
+      update: false,
+      lowSiblingPath: lowSiblingPath,
+      preimage: oldLowWitness,
+      index,
+    };
+
+    return result;
+  }
+}
+
+/****************************************************/
+/****** Some useful Structs and Enums **************/
+/****************************************************/
+enum TreeType {
+  LEAF,
+  BRANCH,
+  EMPTY,
+}
+
+type Leaf = {
+  tag: TreeType.LEAF;
+  value: Fr;
+};
+type Branch = {
+  tag: TreeType.BRANCH;
+  leftTree: Tree;
+  rightTree: Tree;
+};
+type Empty = {
+  tag: TreeType.EMPTY;
+};
+
+type Tree = Leaf | Branch | Empty;
+
+enum SiblingStatus {
+  MEMBER,
+  NONMEMBER,
+  ERROR,
+}
+
+type AccumulatedSiblingPath = {
+  path: Fr[];
+  status: SiblingStatus;
+};
+
+/****************************************************/
+/****** Some Helpful Constructors for Trees  ********/
+/****************************************************/
+const Branch = (left: Tree, right: Tree): Branch => ({
+  tag: TreeType.BRANCH,
+  leftTree: left,
+  rightTree: right,
+});
```

Comment:
IMO Node is a better name than Branch here. I don't think of "branch" as having left and right node, I think of branch as the left _or_ the right path.

Reply:
The main difference is that there is no value in this struct, typically nodes have an associated value.